File: shells/tvos/TinyTrackParty/Render/NameTags.swift
import QuartzCore
import UIKit

/// Who a name tag names: what the sticker shows, and nothing about where.
struct NameTagLabel: Hashable {
    let name: String
    let colorIndex: Int
}

/// The name tags over every other car, the tvOS half of `Stage._paintNameTags`.
///
/// **THE ONE PER-FRAME CHROME, and it is not SwiftUI.** A tag rides a moving car,
/// so its position changes every frame; re-laying a SwiftUI tree at 60 Hz is the
/// churn `DisplayHost` exists to avoid. This is a plain view of pooled layers
/// whose images are drawn ONCE per name and colour, so a frame only moves
/// layers: position, transform, opacity — Core Animation properties, no layout
/// and no text shaping.
///
/// A SUBVIEW OF THE METAL SURFACE, which is what puts it under the SwiftUI cell
/// chips for free: they are the surface's siblings, above it in the ZStack.
///
/// Where each tag goes is C++'s (`ttp/name_tags.h`). Nothing here syncs with the
/// Metal present: the update lands in the same main-thread tick as the frame it
/// was read off, and that is the whole of it (user decision — simple, and never
/// blocking the loop).
final class NameTagView: UIView {

    /// `.name-tag`'s `clamp(1rem, 1.3vw, 1.5rem)` against a 1920-wide display:
    /// 1.3vw is 25, so the clamp lands on its 1.5rem ceiling — the same reading
    /// `NameChip` takes of its own clamp (2.2rem there, 35 pt).
    private static let fontSize: CGFloat = 24

    /// The tag's tilt, `rotate(-2deg)`.
    private static let tilt: CGFloat = -2 * .pi / 180

    private struct TagImage {
        let image: CGImage
        let size: CGSize
        /// The projected point, in the layer's unit coordinates: below the tail's
        /// tip, which is why it is past 1.
        let anchor: CGPoint
    }

    private var stickers: [NameTagLabel: TagImage] = [:]
    private var pool: [CALayer] = []
    private var shown: [NameTagLabel?] = []
    private var visible = 0

    override init(frame: CGRect) {
        super.init(frame: frame)
        isUserInteractionEnabled = false
        backgroundColor = .clear
        autoresizingMask = [.flexibleWidth, .flexibleHeight]
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError("NameTagView is created in code, by DisplayHost") }

    /// Drop sticker images for labels no longer in the field, so a long party of
    /// renames does not keep every old name's bitmap.
    func keepImages(for labels: [NameTagLabel]) {
        stickers = stickers.filter { labels.contains($0.key) }
    }

    /// Place this frame's tags. `packed` is `ttp_display_name_tags`' answer,
    /// `count` tags of 6 floats; `labels` is indexed by roster slot.
    func place(_ packed: UnsafeBufferPointer<Float>, count: Int, labels: [NameTagLabel?]) {
        let w = bounds.width, h = bounds.height
        CATransaction.begin()
        CATransaction.setDisableActions(true)
        var n = 0
        for i in 0..<count where w > 0 && h > 0 {
            let o = i * 6
            let slot = Int(packed[o + 1])
            guard slot >= 0, slot < labels.count, let label = labels[slot] else { continue }
            let layer = n < pool.count ? pool[n] : grow()
            if n >= shown.count || shown[n] != label {
                let s = sticker(for: label)
                layer.contents = s.image
                layer.bounds = CGRect(origin: .zero, size: s.size)
                layer.anchorPoint = s.anchor
                if n < shown.count { shown[n] = label } else { shown.append(label) }
            }
            layer.position = CGPoint(x: CGFloat(packed[o + 2]) * w, y: CGFloat(packed[o + 3]) * h)
            let k = CGFloat(packed[o + 4])
            layer.setAffineTransform(CGAffineTransform(scaleX: k, y: k).rotated(by: Self.tilt))
            layer.opacity = packed[o + 5]
            if n >= visible { layer.isHidden = false }
            n += 1
        }
        for j in n..<max(n, visible) { pool[j].isHidden = true }
        visible = n
        CATransaction.commit()
    }

    /// Hide every tag: no scene, or no race cells.
    func clear() {
        guard visible > 0 else { return }
        CATransaction.begin()
        CATransaction.setDisableActions(true)
        for j in 0..<visible { pool[j].isHidden = true }
        visible = 0
        CATransaction.commit()
    }

    private func grow() -> CALayer {
        let l = CALayer()
        l.contentsScale = window?.screen.scale ?? UIScreen.main.scale
        l.isHidden = true
        // In paint order: C++ lists a cell's tags far to near, so a later layer
        // is a nearer car and belongs on top.
        layer.addSublayer(l)
        pool.append(l)
        return l
    }

    private func sticker(for label: NameTagLabel) -> TagImage {
        if let s = stickers[label] { return s }
        let s = Self.draw(label)
        stickers[label] = s
        return s
    }

    /// `.name-tag` and its tail, drawn once. The box is the CSS box: `padding:
    /// 0.25em 0.5em` around a line-height-1 line, inside the sticker outline,
    /// with `--r-sm` corners and NO shadow. The tail is the web's two triangles:
    /// an ink one hung from the inner edge of the bottom outline, then the livery
    /// one lifted by the outline times sqrt(2), which on 45-degree sides is one
    /// outline width, so the ink runs unbroken round the tip.
    private static func draw(_ label: NameTagLabel) -> TagImage {
        let em = fontSize
        let border = Sticker.border
        let radius = Sticker.radiusSmall
        let font = Fonts.uiDisplay(em, weight: .bold)
        let attrs: [NSAttributedString.Key: Any] = [.font: font, .foregroundColor: UIColor.white]
        let text = label.name as NSString
        let textW = ceil(text.size(withAttributes: attrs).width)

        let boxW = textW + em + 2 * border
        let boxH = em * 1.5 + 2 * border
        let tailH = 0.45 * em + border
        let tailW = 2 * tailH
        let canvas = CGSize(width: boxW, height: boxH - border + tailH)

        let ink = Tokens.uiColor("ink")
        let livery = Tokens.uiCar(label.colorIndex)
        let format = UIGraphicsImageRendererFormat()
        format.scale = UIScreen.main.scale
        format.opaque = false
        let image = UIGraphicsImageRenderer(size: canvas, format: format).image { _ in
            let box = CGRect(x: 0, y: 0, width: boxW, height: boxH)
            ink.setFill()
            UIBezierPath(roundedRect: box, cornerRadius: radius).fill()
            livery.setFill()
            UIBezierPath(roundedRect: box.insetBy(dx: border, dy: border),
                         cornerRadius: max(0, radius - border)).fill()

            func tail(top: CGFloat, color: UIColor) {
                let x0 = (boxW - tailW) / 2
                let p = UIBezierPath()
                p.move(to: CGPoint(x: x0, y: top))
                p.addLine(to: CGPoint(x: x0 + tailW, y: top))
                p.addLine(to: CGPoint(x: x0 + 0.54 * tailW, y: top + 0.92 * tailH))
                p.addLine(to: CGPoint(x: x0 + 0.46 * tailW, y: top + 0.92 * tailH))
                p.close()
                color.setFill()
                p.fill()
            }
            tail(top: boxH - border, color: ink)
            tail(top: boxH - border - border * 2.squareRoot(), color: livery)

            // Centred on the line box the CSS lays out: one em tall, the glyphs'
            // ascender-to-descender span centred inside it.
            let lineTop = border + 0.25 * em
            let glyphH = font.ascender - font.descender
            text.draw(at: CGPoint(x: border + 0.5 * em, y: lineTop + (em - glyphH) / 2), withAttributes: attrs)
        }
        // Stage._paintNameTags: the box's bottom sits 0.55em above the projected
        // point, which leaves the tail's tip just clear of the car.
        let anchorY = (boxH + 0.55 * em) / canvas.height
        return TagImage(image: image.cgImage!, size: canvas, anchor: CGPoint(x: 0.5, y: anchorY))
    }
}

File: shells/androidtv/app/src/main/kotlin/games/couchpad/tinytrack/NameTags.kt
package games.couchpad.tinytrack

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Typeface
import android.view.View
import androidx.compose.ui.graphics.toArgb
import kotlin.math.ceil
import kotlin.math.sqrt

/**
 * The name tags over every other car in each cell, CPU included — the Android
 * half of the web's `Stage._paintNameTags` and its `.name-tag` CSS.
 *
 * **THE ONE HUD ELEMENT PLACED PER FRAME, AND WHY IT IS NOT COMPOSE.** A tag
 * rides a moving car, so it moves on every presented frame, and Compose shares
 * this thread with the renderer: a recomposition that reaches layout costs 10-30
 * ms on the box (`shells/androidtv/CLAUDE.md`). This is a plain View for the
 * reason [PerfOverlayView] is one. WHERE each tag goes is C++'s
 * (`ttp_display_name_tags`, read by [DisplayHost] right after the frame it
 * describes); this decides nothing.
 *
 * **A STICKER IS DRAWN ONCE, NOT PER FRAME.** Each car's tag — livery fill, ink
 * outline, tail, white Fredoka — is rasterized into a Bitmap when its name or
 * colour changes, and a frame only draws those bitmaps through a matrix. No text
 * is shaped and nothing is allocated on the frame path.
 *
 * Sits UNDER the ComposeView, so the cell chips and cards paint over a tag that
 * drifts beneath them, as on the web.
 */
class NameTagView(context: Context) : View(context) {

    private class TagBitmap(val name: String, val color: Int, val bitmap: Bitmap,
                          val anchorX: Float, val anchorY: Float)

    /** C++ writes into this directly ([DisplayHost]); [count] says how much of it is live. */
    val tags = FloatArray(MAX_TAGS * TAG_STRIDE)
    private var count = 0

    /** Slot i's car id, latched per scene build by [DisplayHost]. */
    private var slots: List<EngineId> = emptyList()
    private var stickers = HashMap<EngineId, TagBitmap>()

    /** Authored pixels (1920 wide) to this window's physical pixels. */
    private val k = resources.displayMetrics.widthPixels / AUTHORED_WIDTH
    private val face = Typeface.createFromAsset(context.assets, "fonts/Fredoka-Bold.ttf")
    private val matrix = Matrix()
    private val paint = Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG)

    init {
        isFocusable = false
        setWillNotDraw(false)
    }

    /**
     * Who the field is: re-rasterizes only the stickers whose name or livery
     * moved, and drops the ones whose car left. Called whenever the scene's cars
     * change (a build, a rename), never per frame.
     */
    fun setField(cars: List<SceneCar>) {
        val next = HashMap<EngineId, TagBitmap>(cars.size * 2)
        for (car in cars) {
            val color = Tokens.car(car.colorIndex).toArgb()
            val have = stickers[car.id]
            next[car.id] = if (have != null && have.name == car.name && have.color == color) have
                else rasterize(car.name, color)
        }
        stickers = next
        invalidate()
    }

    /** The scene's slot order, off the built roster (`ttp_display_slot_ids_json`). */
    fun setSlots(ids: List<EngineId>) { slots = ids }

    /** How many tags C++ just wrote into [tags]; 0 hides them all. */
    fun show(n: Int) {
        if (n == 0 && count == 0) return
        count = n
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        val w = width.toFloat()
        val h = height.toFloat()
        for (i in 0 until count) {
            val o = i * TAG_STRIDE
            val slot = tags[o + 1].toInt()
            val s = stickers[slots.getOrNull(slot) ?: continue] ?: continue
            val scale = tags[o + 4]
            matrix.setTranslate(-s.anchorX, -s.anchorY)
            matrix.postRotate(TILT)
            matrix.postScale(scale, scale)
            matrix.postTranslate(tags[o + 2] * w, tags[o + 3] * h)
            paint.alpha = (tags[o + 5] * 255f).toInt().coerceIn(0, 255)
            canvas.drawBitmap(s.bitmap, matrix, paint)
        }
    }

    /**
     * One tag, at full size: `.name-tag` plus its `::before`/`::after`. The anchor
     * is where the car's projected point goes — the tail's tip sits just above
     * it, because the box's bottom edge is [GAP_EM] above.
     */
    private fun rasterize(name: String, color: Int): TagBitmap {
        val em = FONT_PX * k
        val border = Sticker.border.value * k
        val radius = Tokens.radius("r-sm").value * k
        val text = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            typeface = face
            textSize = em
            this.color = android.graphics.Color.WHITE
        }
        // `line-height: 1` and `padding: 0.25em 0.5em`, border-box.
        val boxW = ceil(text.measureText(name) + em + 2 * border)
        val boxH = ceil(em * 1.5f + 2 * border)
        val tailH = 0.45f * em + border
        val tailW = 2 * tailH
        val bmpH = ceil(boxH + tailH)
        val bitmap = Bitmap.createBitmap(boxW.toInt(), bmpH.toInt(), Bitmap.Config.ARGB_8888)
        val c = Canvas(bitmap)
        val fill = Paint(Paint.ANTI_ALIAS_FLAG)

        // The ink outline is the box itself, and the livery an inset of it with
        // the inner radius CSS gives a border-box corner.
        fill.color = Tokens.ink.toArgb()
        c.drawRoundRect(RectF(0f, 0f, boxW, boxH), radius, radius, fill)
        fill.color = color
        val inner = (radius - border).coerceAtLeast(0f)
        c.drawRoundRect(RectF(border, border, boxW - border, boxH - border), inner, inner, fill)

        // THE TAIL: an ink triangle hung from the INNER edge of the bottom outline,
        // then the same triangle in the livery lifted by border * sqrt(2). On
        // 45-degree sides that lift is exactly one border perpendicular to each
        // side, so the ink runs as one band from the box round the tip. The flat
        // tip is the web's soft point.
        val top = boxH - border
        fill.color = Tokens.ink.toArgb()
        c.drawPath(tail(boxW / 2, top, tailW, tailH), fill)
        fill.color = color
        c.drawPath(tail(boxW / 2, top - border * sqrt(2f), tailW, tailH), fill)

        // Centred on the content box by the face's own metrics, so it sits where
        // the browser's line box puts it.
        val fm = text.fontMetrics
        val baseline = boxH / 2 - (fm.ascent + fm.descent) / 2
        c.drawText(name, border + em * 0.5f, baseline, text)

        return TagBitmap(name, color, bitmap, boxW / 2, boxH + GAP_EM * em)
    }

    private fun tail(cx: Float, top: Float, w: Float, h: Float) = Path().apply {
        moveTo(cx - w / 2, top)
        lineTo(cx + w / 2, top)
        lineTo(cx + w * 0.04f, top + h * 0.92f)
        lineTo(cx - w * 0.04f, top + h * 0.92f)
        close()
    }

    private companion object {
        /**
         * `clamp(1rem, 1.3vw, 1.5rem)` on a 1920-wide display lands on its 24 px
         * ceiling, against the name chip's 35 — the ratio this shell's own 35 px
         * chip keeps ([RaceHud]'s NameChip).
         */
        const val FONT_PX = 24f
        /** The box's bottom edge above the anchor, in em (`translateY(-0.55em)`). */
        const val GAP_EM = 0.55f
        const val TILT = -2f
        const val TAG_STRIDE = 6
        /** Eight cars, so seven tags a cell, across the most cells a field can split into. */
        const val MAX_TAGS = 8 * 7
    }
}
